C++23 deprecates std::aligned_storage and std::aligned_storage_t

<!--
PLEASE DO NOT POST A QUESTION HERE.
This form is for bug reports and feature requests ONLY!

For general questions and troubleshooting, please ask/look for answers at StackOverflow, with "grpc" tag: https://stackoverflow.com/questions/tagged/grpc

For questions that specifically need to be answered by gRPC team members, please ask/look for answers at grpc.io mailing list: https://groups.google.com/forum/#!forum/grpc-io

Issues specific to *grpc-java*, *grpc-go*, *grpc-node*, *grpc-dart*, *grpc-web* should be created in the repository they belong to (e.g. https://github.com/grpc/grpc-LANGUAGE/issues/new)
-->
C++23 deprecates std::aligned_storage and std::aligned_storage_t. This should be addressed in upcoming versions. :)


### What version of gRPC and what language are you using?
gRPC 1.52.1

### What operating system (Linux, Windows,...) and version?
Windows Server 2016
RHEL 8.7

### What runtime / compiler are you using (e.g. python version or version of gcc)
Visual Studio  2019 (16.11.25)
clang 15.0.7

### What did you do?
Please provide either 1) A unit test for reproducing the bug or 2) Specific steps for us to follow to reproduce the bug. If there’s not enough information to debug the problem, gRPC team may close the issue at their discretion. You’re welcome to re-open the issue once you have a reproduction.
Set language standard to c++23

### What did you expect to see?
Compiles without warnings.

### What did you see instead?
Windows:
`gRPC\1.52.1\include\grpcpp/server_context.h(538): warning C4996: 'std::aligned_storage<104,8>::type': warning STL4034: std::aligned_storage and std::aligned_storage_t are deprecated in C++23. Prefer alignas(T) std::byte t_buff[sizeof(T)]. You can define _SILENCE_CXX23_ALIGNED_STORAGE_DEPRECATION_WARNING or _SILENCE_ALL_CXX23_DEPRECATION_WARNINGS to suppress this warning.
`

Linux:
`gRPC/1.52.1/include/grpcpp/server_context.h:538:17: warning: 'aligned_storage<88, 8>' is deprecated [-Wdeprecated-declarations]
`

Make sure you include information that can help us debug (full error message, exception listing, stack trace, logs).

See [TROUBLESHOOTING.md](https://github.com/grpc/grpc/blob/master/TROUBLESHOOTING.md) for how to diagnose problems better.

### Anything else we should know about your project / environment?

--- a/include/grpcpp/server_context.h
+++ b/include/grpcpp/server_context.h
@@ -545,8 +545,7 @@ class ServerContextBase {
     const std::function<void(grpc::Status s)> func_;
   };
 
-  typename std::aligned_storage<sizeof(Reactor), alignof(Reactor)>::type
-      default_reactor_;
+  alignas(Reactor) char default_reactor_[sizeof(Reactor)];
   std::atomic_bool default_reactor_used_{false};
 
   std::atomic_bool marked_cancelled_{false};

--- a/src/core/lib/gprpp/manual_constructor.h
+++ b/src/core/lib/gprpp/manual_constructor.h
@@ -25,7 +25,6 @@
 
 #include <stddef.h>
 
-#include <type_traits>
 #include <utility>
 
 #include "src/core/lib/gprpp/construct_destruct.h"
@@ -139,7 +138,7 @@ class ManualConstructor {
   void Destroy() { Destruct(get()); }
 
  private:
-  typename std::aligned_storage<sizeof(Type), alignof(Type)>::type space_;
+  alignas(Type) char space_[sizeof(Type)];
 };
 
 }  // namespace grpc_core

--- a/src/core/lib/gprpp/no_destruct.h
+++ b/src/core/lib/gprpp/no_destruct.h
@@ -68,7 +68,7 @@ class NoDestruct {
   const T* get() const { return reinterpret_cast<const T*>(&space_); }
 
  private:
-  typename std::aligned_storage<sizeof(T), alignof(T)>::type space_;
+  alignas(T) char space_[sizeof(T)];
 };
 
 // Helper for when a program desires a single *process wide* instance of a

--- a/src/core/lib/iomgr/event_engine_shims/endpoint.cc
+++ b/src/core/lib/iomgr/event_engine_shims/endpoint.cc
@@ -60,10 +60,8 @@ class EventEngineEndpointWrapper {
   struct grpc_event_engine_endpoint {
     grpc_endpoint base;
     EventEngineEndpointWrapper* wrapper;
-    std::aligned_storage<sizeof(SliceBuffer), alignof(SliceBuffer)>::type
-        read_buffer;
-    std::aligned_storage<sizeof(SliceBuffer), alignof(SliceBuffer)>::type
-        write_buffer;
+    alignas(SliceBuffer) char read_buffer[sizeof(SliceBuffer)];
+    alignas(SliceBuffer) char write_buffer[sizeof(SliceBuffer)];
   };
 
   explicit EventEngineEndpointWrapper(

--- a/src/core/lib/promise/arena_promise.h
+++ b/src/core/lib/promise/arena_promise.h
@@ -19,6 +19,7 @@
 
 #include <stdlib.h>
 
+#include <cstddef>
 #include <memory>
 #include <type_traits>
 #include <utility>
@@ -34,7 +35,10 @@ namespace grpc_core {
 
 namespace arena_promise_detail {
 
-using ArgType = std::aligned_storage_t<sizeof(void*)>;
+struct ArgType {
+  alignas(std::max_align_t) char buffer[sizeof(void*)];
+};
+
 template <typename T>
 T*& ArgAsPtr(ArgType* arg) {
   static_assert(sizeof(ArgType) >= sizeof(T**),

--- a/test/core/gprpp/ref_counted_test.cc
+++ b/test/core/gprpp/ref_counted_test.cc
@@ -119,8 +119,8 @@ class ValueInExternalAllocation
 };
 
 TEST(RefCounted, CallDtorUponUnref) {
-  std::aligned_storage<sizeof(ValueInExternalAllocation),
-                       alignof(ValueInExternalAllocation)>::type storage;
+  alignas(ValueInExternalAllocation) char
+      storage[sizeof(ValueInExternalAllocation)];
   RefCountedPtr<ValueInExternalAllocation> value(
       new (&storage) ValueInExternalAllocation(5));
   EXPECT_EQ(value->value(), 5);

